### ISSUE TITLE
Profile, mem-cache, refactor

### ISSUE THREAD

**Issue by ddrake** [ddrake/Integrated_Farm_Budget_Tool#142]
Profile the data collection methods for detailed budget and sensitivity.  Cache values within a single response and return cached values when available.  Look for large methods and refactor.  Add one or more sensitivity table tests beforehand.

**Comment by ddrake** [comment]
I wasn't able to get profiling to work just using the built-in Python cProfiler because of nested module complexities.  There is a [Django profiler](https://pypi.org/project/django-cprofile-middleware/1.0/) that would be reasonable to try, but I haven't looked into it much yet.  

Knowing  the code pretty well, I tried caching the results of frequently-called functions in memory such that the cache is automatically cleared after each request.  This gave about a 2.8% speed up for the budget and about an 11.1% speed up for sensitivity (averaged over 20 trials).

There is some risk to this, in that if one of the functions whose result is cached were to be called with different arguments in the same request (not currently the case), the resulting error could be hard to track down.

I'm going to merge in the mem-cached branch for now, since it also contains bug fixes.  The caching can easily be removed later if we like.

**Comment by ddrake** [comment]
I installed [django-cproflile-middleware](https://github.com/omarish/django-cprofile-middleware), but it gave me an error.  Luckily the fix was present as an open pull request.  Here's [what I did](https://github.com/omarish/django-cprofile-middleware/issues/30#issuecomment-1739678041) to get it working.

I ran it on the sensitivity table like this: http://127.0.0.1:8000/sensitivity/19?prof&sort=ncalls&count=500

That gave me some useful information on how many times some methods were called and the time used:
```
      ncalls  tottime  percall  cumtime  percall filename:lineno(function) 
       20    0.000    0.000    0.000    0.000 indemnity.py:323(revenue_trigger_feb_price_area)
       19    0.000    0.000    0.000    0.000 market_crop.py:32(__init__)
       18    0.000    0.000    0.022    0.001 sens_table.py:706(__init__)
       17    0.000    0.000    0.000    0.000 market_crop.py:50()
       17    0.001    0.000    0.001    0.000 {method 'fetchone' of 'psycopg2.extensions.cursor' objects}
       16    0.000    0.000    0.001    0.000 sens_table.py:527(__init__)
       16    0.000    0.000    0.021    0.001 sens_table.py:613(__init__)
       16    0.000    0.000    0.019    0.001 util.py:19(any_changed)
       16    0.000    0.000    0.000    0.000 indemnity.py:155(rev_trigger_condition)
       16    0.000    0.000    0.001    0.000 indemnity.py:422(county_rev_as_ratio)
       16    0.000    0.000    0.000    0.000 indemnity.py:440(actual_revenue_opt)
       16    0.000    0.000    0.000    0.000 indemnity.py:458(county_insured_revenue)
       16    0.000    0.000    0.000    0.000 market_crop.py:56()
       16    0.000    0.000    0.014    0.001 market_crop.py:142(expected_total_bushels)
       16    0.000    0.000    0.014    0.001 market_crop.py:157(production_frac_for_farm_crop)
       16    0.000    0.000    0.000    0.000 farm_crop.py:175(get_item)
       16    0.000    0.000    0.000    0.000 farm_crop.py:300()
       16    0.000    0.000    0.000    0.000 farm_crop.py:384() 
```

This shows that it could make sense to cache the results of `expected_total_bushels`  and `production_frac_for_farm_crop` in the `MarketCrop` model.  Also the `any_changed` function in util.py is more expensive than I thought.  Why is it getting called 16 times?

**Comment by ddrake** [comment]
The answer to the last question:   To handle the fact that the insurance settings for fs beans and dc beans have to stay the same, I added a check in the farm_crop.save() method to see if e.g. the coverage level changed so if it did we could update the related crop.  That check was causing the (expensive) method `any_changed` to be called whenever we saved a farm crop.  

The reason farm crops are getting saved when the sensitivity table is generated is that, in order to have premiums available after the farm year has changed, we save all the premiums (for every choice) each time the premiums are computed.  Because of complexities around time travel, we recompute premiums whenever they are requested (unless we are in a past farm year, in which case we can't compute them because we no longer have the actuarial data).

To eliminate the expensive any_changed calls, I made a small modification so that when we're saving the farm_crop just because premiums are requested, we don't bother with the check for a related crop changing!
